File: front-end/Mobile/src/models/attendance/Justification.js
import BaseModel from '../BaseModel';

export default class Justification extends BaseModel {
  constructor(data = {}) {
    super(data);
    this.justificationId = data.justification_id || null;
    this.attendanceRecordId = data.attendance_record_id || null;
    this.justificationTypeId = data.justification_type_id || null;
    this.reason = data.reason || '';
    this.submittedAt = data.submitted_at || null;
    this.reviewedBy = data.reviewed_by || null;
    this.reviewedAt = data.reviewed_at || null;
    this.reviewStatus = data.review_status || 'Pending';
  }

  get isPending() {
    return this.reviewStatus === 'Pending';
  }

  get isApproved() {
    return this.reviewStatus === 'Approved';
  }

  get isRejected() {
    return this.reviewStatus === 'Rejected';
  }

  get statusColor() {
    const colors = {
      Pending: '#FF9800',
      Approved: '#2da351',
      Rejected: '#ff0000',
    };
    return colors[this.reviewStatus] || '#999999';
  }

  static fromApi(data) {
    if (!data) return null;
    return new Justification(data);
  }

  toApi() {
    return {
      ...super.toApi(),
      justification_id: this.justificationId,
      attendance_record_id: this.attendanceRecordId,
      justification_type_id: this.justificationTypeId,
      reason: this.reason,
      submitted_at: this.submittedAt,
      reviewed_by: this.reviewedBy,
      reviewed_at: this.reviewedAt,
      review_status: this.reviewStatus,
    };
  }
}

File: back-end/05-ms-attendance/src/main/java/com/faceattend_edu/attendance_service/domain/model/Justification.java
package com.faceattend_edu.attendance_service.domain.model;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.Instant;
import java.util.Objects;
import java.util.Set;
import java.util.UUID;

@Getter @Setter @AllArgsConstructor @NoArgsConstructor
public class Justification {
    private Long justificationId;
    private Long attendanceRecordId;
    private Integer justificationTypeId;
    private String reason;
    private Instant submittedAt;
    private UUID reviewedBy;
    private Instant reviewedAt;
    private String reviewStatus;
    private String resolutionNotes;
    private Instant createdAt;
    private Instant updatedAt;
    private Instant deletedAt;
    private UUID createdBy;
    private UUID updatedBy;
    private UUID deletedBy;
    private long rowVersion;

    private static final Set<String> ALLOWED = Set.of("Pending","Approved","Rejected");

    public void validate() {
        if (attendanceRecordId == null) throw new IllegalArgumentException("attendanceRecordId is required");
        if (justificationTypeId == null) throw new IllegalArgumentException("justificationTypeId is required");
        if (isBlank(reason)) throw new IllegalArgumentException("reason is required");
        if (reviewStatus != null && !ALLOWED.contains(reviewStatus)) throw new IllegalArgumentException("reviewStatus must be one of " + ALLOWED);
    }
    public void touchCreated() {
        if (createdAt == null) createdAt = Instant.now();
        if (submittedAt == null) submittedAt = Instant.now();
        if (reviewStatus == null) reviewStatus = "Pending";
        if (rowVersion == 0) rowVersion = 1L;
    }
    public void touchUpdated(){ updatedAt = Instant.now(); }
    private static boolean isBlank(String s){ return s==null || s.trim().isEmpty(); }
    @Override public boolean equals(Object o){ if(this==o) return true; if(!(o instanceof Justification)) return false; Justification that=(Justification)o; return Objects.equals(justificationId, that.justificationId); }
    @Override public int hashCode(){ return Objects.hash(justificationId); }
}
